pylabrobot/pumps/agrowpumps/agrowdosepump.py

```diff
@@ -0,0 +1,130 @@
+import threading
+import time
+from typing import Optional, List, Dict
+
+from pylabrobot.pumps.backend import PumpArrayBackend
+
+from pymodbus.client import AsyncModbusSerialClient as ModbusClient  # type: ignore
+
+
+class AgrowPumpArray(PumpArrayBackend):
+  """
+  AgrowPumpArray is a class that allows users to control AgrowPumps via Modbus communication.
+
+  Attributes:
+      port: The port that the AgrowPumpArray is connected to.
+      unit: The unit number of the AgrowPumpArray.
+      pump_index_to_address: A dictionary that maps pump indices to their Modbus addresses.
+
+  Properties:
+      num_channels: The number of channels that the AgrowPumpArray has.
+  """
+
+  def __init__(self, port: str,
+               unit: int,
+               pump_index_to_address: Optional[Dict[int, int]] = None):
+    self.port = port
+    self.unit = unit
+    self.pump_index_to_address = pump_index_to_address
+    self.modbus: Optional[ModbusClient] = None
+
+  @property
+  def num_channels(self) -> int:
+    """
+    num_channels(self): This method returns the number of channels that the AgrowPumpArray has.
+
+    Returns:
+        int: The number of channels that the AgrowPumpArray has.
+
+    """
+    if self.pump_index_to_address is None:
+      if self.modbus.connected:  # type: ignore[union-attr]
+        register_return = \
+          self.modbus.read_holding_registers(19, 2, unit=self.unit)  # type: ignore[union-attr]
+        return int("".join(chr(r // 256) + chr(r % 256) for r in register_return.registers)[2])
+      else:
+        return 0
+    else:
+      return len(self.pump_index_to_address)
+
+  def start_keep_alive_thread(self):
+
+    """
+
+    keep_alive_thread(self): This method creates a daemon thread that sends a Modbus request
+    every 25 seconds to keep the connection alive.
+
+    """
+
+    def keep_alive():
+      """
+      keep_alive(self): This method sends a Modbus request every 25 seconds to keep the
+      connection alive.
+      """
+      while True:
+        time.sleep(25)
+
+        self.modbus.read_holding_registers(0, 1, unit=self.unit)  # type: ignore[union-attr]
+
+    thread = threading.Thread(target=keep_alive, daemon=True)
+    thread.start()
```

Comment:
I think we should maintain a reference to this thread on self so we can terminate it on `stop`.

Reply:
I made this change in the most recent commit